### PR TITLE
add Comparable interface to Dn and Component

### PR DESCRIPTION
**Pull request by frett**

### CODE REVIEW

File: api/src/main/java/org/ccci/idm/user/Dn.java [hunk 1]
```diff
@@ -82,6 +82,19 @@ public final Dn parent() {
     }
 
     @Override
+    public int compareTo(@Nonnull final Dn o) {
```
Comment (frett) [inline api/src/main/java/org/ccci/idm/user/Dn.java]: sort by components in order, then by length if base components are the same

File: api/src/main/java/org/ccci/idm/user/Dn.java [hunk 2]
```diff
@@ -112,6 +125,15 @@ public Component(@Nonnull final String type, @Nonnull final String value) {
         }
 
         @Override
+        public int compareTo(@Nonnull final Component o) {
```
Comment (frett) [inline api/src/main/java/org/ccci/idm/user/Dn.java]: sort by `type`, then `value` case insensitively. I could be swayed to look at `value` before `type`.

that would change this ordering: `cn=Zebra ou=Aardvark` to `ou=Aardvark cn=Zebra`

Comment (hlbraddock) [inline api/src/main/java/org/ccci/idm/user/Dn.java]: Other than just sorting on component value, this PR looks good.

+1

**Comment by frett** [comment]
Updated ordering, will merge once travis completes

File: api/src/main/java/org/ccci/idm/user/Dn.java [hunk 3]
```diff
@@ -112,6 +125,15 @@ public Component(@Nonnull final String type, @Nonnull final String value) {
...
+            int resp = value.compareToIgnoreCase(o.value);
+            if (resp == 0) {
+                resp = type.compareToIgnoreCase(o.type);
+            }
+            return resp;
+        }
```
Comment (hlbraddock) [inline api/src/main/java/org/ccci/idm/user/Dn.java]: Barring this:
```
return value.compareToIgnoreCase(o.value) == 0 ? type.compareToIgnoreCase(o.type) : value.compareToIgnoreCase(o.value);
```

You could:
```
int resp = value.compareToIgnoreCase(o.value);
return resp == 0 ? type.compareToIgnoreCase(o.type) : resp;
```

Comment (frett) [inline api/src/main/java/org/ccci/idm/user/Dn.java]: true, I could have inlined part of it using the ternary operator (also know as the elvis operator `?:`)

I'll leave it as is, someone can refactor into that if they feel more drawn to it in this use case :)

**Comment by waronlive** [comment]
would like some back

**Pull request merged**